### DISCUSSION TITLE
How to set tab as space

### DISCUSSION THREAD

**Discussion by jkmpariab** [helix-editor/helix#2891] [Q&A] [answered]
Is there any way to customize tab size just as `:indent-style INTEGER` does in command console on config/languages file?

[Documentation](https://docs.helix-editor.com/guides/adding_languages.html) suggest setting `indent` key for `[[language]]` as this:
```
indent = { tab-width = INTEGER, unit = "STRING" }
```
By playing some time found out the `unit` field does not have any affect on tab behaviour.
and `tab-width` is the one that just set tab size as its name implies.

BTW by setting `tab-width = 0 ` helix do panic

**Comment by kirawi** [comment, marked as answer]
`tab-width` changes the size of the rendered tab (`\t`), while `unit` is what is inserted when you try to indent.
e.g.
`unit = "\t"` will insert a tab
`unit = "    "` will insert 4 spaces (github is rendering it weird, but there are four space characters inside the quotes)

**Reply by jkmpariab** [reply]
Interestingly it does not work for me.
Do i missed something?
This is my `languages.toml`:
```
[[language]]
name = "go"
indent = { unit = "    " }
```
or even this:
```
[[language]]
name = "go"
indent = { tab-width = 4, unit = "    " }
```

helix --version: `helix 22.05 (27609f50)`

**Reply by kirawi** [reply]
What happens when you try to indent?

**Reply by jkmpariab** [reply]
Ok know i found out there's a bug or  a feature (that i never expect) in how indentation works in helix:

- Indentation at the start of the line is not `unit`-inserted but `tab-width` is ok.
- Yeah at the middle or end of the line it works.

Is this the intended behavior?

**Reply by jkmpariab** [reply]
> Ok know i found out there's a bug or a feature (that i never expect) in how indentation works in helix:
> 
> * Indentation at the start of the line is not `unit`-inserted but `tab-width` is ok.
> * Yeah at the middle or end of the line it works.
> 
> Is this the intended behavior?

I mean it's definitely a bug as indentation at the start of the line is the main use case for me

**Reply by kirawi** [reply]
I think that you're not configuring your `languages.toml` correctly. By default Go will use `\t` for indentation, which is why you're seeing `tab-width` affect indentation (`unit` shouldn't be inserted in the middle or end either, as it only indents the beginning of the line). Where is your `languages.toml` located, and what's your OS?

**Reply by jkmpariab** [reply]
os: `ubuntun 22.04`
config path: `~/.config/helix/languages.toml`
BTW `:indent-style 4` does the same on command console so i think it is not a config issue

**Reply by kirawi** [reply]
Could you post an `asciinema` of what you're seeing? I'm not able to replicate.

**Reply by jkmpariab** [reply]
Ok this is the asciinema:
https://asciinema.org/a/SBF1zI7bMpp2mBSVK76rYd55r

**Reply by kirawi** [reply]
I'm confused, it looks like it's indenting correctly?

Also the reason you're able to insert in the middle of a line is that you're using `tab` instead of `>` to change indentation. `tab` will insert wherever the cursor is, while `>` will change the line's indentation.

**Reply by jkmpariab** [reply]
Unfortunately im not kidding ://
Is there any debug mode to monitor key pressed and events?

**Reply by kirawi** [reply]
There isn't. I'm still confused as to what you're considering a bug. In the asciinema, it correctly inserts the spaces according to the desired length. Regardless of whether you press `tab` or `>`, it will insert the same characters (just the location is different).

**Reply by jkmpariab** [reply]
On pressing `backspace` it clear all of the spaces?
So that's the bug!

**Reply by kirawi** [reply]
Ah, that's an intended feature to auto-dedent upon `backspace`. If you want to disable it you'll have to delete these lines: https://github.com/helix-editor/helix/blob/94fc41a41920fc705f01637e7902f06a1c32d998/helix-term/src/commands.rs#L2949-L2992
and compile.

You can create an issue for a configuration option for it.

**Reply by jkmpariab** [reply]
Ok i found out that the behavior is normal as on vscode too but not on gnome gedit.
So i think that is not a big deal as i thought before.
Thanks for your help!

**Reply by goyalyashpal** [reply]
> On pressing `backspace` it clear all of the spaces?
> So that's the bug!

that is why u (should) enable "render whitespace" lol.

**Comment by giancarlocp** [comment]
How to set tab behavior globally?

**Reply by A-kly** [reply]
I also want an answer to this, seems to be a simple change to the [editor] section of the TOML file

**Reply by A-kly** [reply]
Hello! Has there been any update to this?

**Reply by jm355** [reply]
```
[[language]]
name = "*"
scope = "*"
file-types = [ { glob = "*" } ]
indent = { tab-width = 4, unit = "    " }
```

See also #15620
